Rationale.md

```diff
@@ -566,3 +566,44 @@ Yes:
 [future general]: FutureFeatures.md
 [future flow control]: FutureFeatures.md#more-expressive-control-flow
 [future integers]: FutureFeatures.md#additional-integer-operations
+
+
+## Why "polymorphic" stack typing after control transfer?
+
+For a stack machine, the fundamental property that type checking must ensure is that for every _edge_ in a control flow graph (formed by the individual instructions), the assumptions about the stack match up.
+At the same time, type checking should be fast, so be expressible by a linear traversal.
+
+There is no control edge from an unconditional control transfer instruction (like `br`, `br_table`, `return`, or `unreachable`) to the textually following instruction -- the next instruction is unreachable.
+Consequently, no constraint has to be imposed between the two regarding the stack.
+In a linear type checking algorithm this can be expressed by allowing to assume any type of stack.
+In type system terms, the instruction can thus be said to be _polymorphic_ in its stack type.
+
+This solution is canonical in the sense that it induces the minimal type constraints necessary to ensure soundness, while also maintaining the following useful structural properties about valid (i.e., well-typed) code:
+
+* It is composable: if statement sequences A and B are valid, then block(A,B) is valid.
+* It is decomposable: if block(A,B) is valid, then both A and B are valid.
```

Comment:
Don't these two statements need some sort of qualification that the signature of A and B are `[]->[]`?